### PR TITLE
Local timestamps

### PR DESCRIPTION
**Pull request by nomeata**
I gave #378 a shot. The code turned out to be not too large, which is good. A very very quick test runs, but the CPU usage was very high (possibly due to missing pings). Review and possibly merging would be nice. I will have less time over the weekend.

### CODE REVIEW

**Comment by nomeata** [comment]
The prediction code as well as the network layer now allows for timestamps to be sent. We just need the driver to create a thread to send them. You have a better idea about the driver and such – would you add that?

File: codeworld-api/src/CodeWorld/Prediction.hs [hunk 1]
```diff
@@ -21,109 +21,144 @@ multi-player setup.
...
-    | now - commitTime > rate
-    = serverEvent step rate ours now event (commitedRateStep step rate f)
+-- Time handling.
+--
+-- Move state forward in fixed animation rate steps, and get
+-- the timestamp as close to the given target as possible (but possibly stop short)
+timePassesBigStep :: StepFun s -> AnimationRate -> Timestamp -> TState s -> TState s
+timePassesBigStep step rate target (now, s)
+    | now + rate < target
+    = timePasses step rate target (stepBy step rate (now, s))
```
Comment (nomeata) [inline codeworld-api/src/CodeWorld/Prediction.hs]: Yes, Typo! Thanks.

**Comment by nomeata** [comment]
I am right now writing a test suite for the prediction code that tries to find out that no matter in what order events come in, the resulting state is the same. That should hopefully give some assurance in this not really trivial code.

**Comment by cdsmith** [comment]
Okay.  Do you want to wait on the tests to merge this pull request?  Or I'm happy to merge it once the typo is fixed, and the test can be committed separately.  Your choice.

**Comment by nomeata** [comment]
Since this is the second bug I had in the code, I’ll finish the test first.

**Comment by nomeata** [comment]
Ok, unsurprisingly, writing a test case uncovered some corner cases. The test suite is not yet comprehensive in any way, but I guess this can be merged.

NB: The clients still don't actually send out pings.

**Comment by cdsmith** [comment]
Looks like a good start.  I'll merge this, and we can iterate from here.
